### ISSUE TITLE
Floating Point Rounding Error using int()

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#3460]
Migrated from [rt.perl.org#5881](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=5881) (status was 'resolved')

Searchable as RT5881$

**Comment by p5pRT** [comment]
### From cdhs@tara.commerce.uk.net



I've found what I think is a floating point rounding error using the
int\(\) function\. I've included four perl snippets\, with their results&#8203;:

&nbsp;  \[1\]     28\.4 \* 1\.175 \* 100              =>       3337
&nbsp;  \[2\]     28\.4 \* 117\.5                    =>       3337
&nbsp;  \[3\]     int\( 28\.4 \* 1\.175 \* 100 \)       =>       3336
&nbsp;  \[4\]     int\( 28\.4 \* 117\.5       \)       =>       3337

I came across this when calculating tax\-inclusive prices @&#8203; 17\.5%

I am now using \[4\] instead of \[3\]\, which doesn't appear to barf on any
values that I throw at it\, but I'd like to know exactly what is
happening here\.

I assume that the result of the arithmetic above is 3336\.999999999
or similar\, and it is rounded by perl because of precision differences\,
but in \[3\] int\(\) is given the un\-rounded value and returning the integer
part\. What I don't understand is why \[3\] differs to \[4\]\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.6.0:

Configured by flepied at Sat Sep 30 18:19:36 CEST 2000.

Summary of my perl5 (revision 5.0 version 6 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.2.17-16mdksmp, archname=i386-linux
    uname='linux debris.mandrakesoft.com 2.2.17-16mdksmp #1 smp thu sep 28 15:25:49 cest 2000 i686 unknown '
    config_args='-des -Darchname=i386-linux -Dd_dosuid -Ud_csh -Duseshrplib -Doptimize=-O3 -fomit-frame-pointer -pipe -s -march=pentium -ffast-math -fexpensive-optimizations -Dprefix=/usr -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/lib/perl5/man/man3'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define 
    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
  Compiler:
    cc='cc', optimize='-O3 -fomit-frame-pointer -pipe -s -march=pentium -ffast-math -fexpensive-optimizations', gccversion=2.95.3 19991030 (prerelease)
    cppflags='-fno-strict-aliasing'
    ccflags ='-fno-strict-aliasing -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    stdchar='char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -ldl -lm -lc -lposix -lcrypt
    libc=/lib/libc-2.1.3.so, so=so, useshrplib=true, libperl=libperl.so
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic -Wl,-rpath,/usr/lib/perl5/5.6.0/i386-linux/CORE'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.6.0:
    /usr/lib/perl5/5.6.0/i386-linux
    /usr/lib/perl5/5.6.0
    /usr/lib/perl5/site_perl/5.6.0/i386-linux
    /usr/lib/perl5/site_perl/5.6.0
    /usr/lib/perl5/site_perl
    .


Environment for perl v5.6.0:
    HOME=/home/cdhs
    LANG=en
    LANGUAGE=en_GB:en
    LC_COLLATE=en_GB
    LC_CTYPE=en_GB
    LC_MESSAGES=en_GB
    LC_MONETARY=en_GB
    LC_NUMERIC=en_GB
    LC_TIME=en_GB
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/local/bin:/bin:/usr/bin:/usr/X11R6/bin:/home/cdhs/bin:/usr/X11R6/bin:/home/cdhs/bin:/usr/X11R6/bin:/usr/X11R6/bin:/usr/X11R6/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @nwc10

On Wed\, Feb 21\, 2001 at 05&#8203;:49&#8203;:02PM \+0000\, cdhs@&#8203;tara\.commerce\.uk\.net wrote&#8203;:
> 
> This is a bug report for perl from cdhs@&#8203;commerce\.uk\.net\,
> generated with the help of perlbug 1\.28 running under perl v5\.6\.0\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> I've found what I think is a floating point rounding error using the
> int\(\) function\. I've included four perl snippets\, with their results&#8203;:
> 
>   \[1\]     28\.4 \* 1\.175 \* 100              =>       3337
>   \[2\]     28\.4 \* 117\.5                    =>       3337
>   \[3\]     int\( 28\.4 \* 1\.175 \* 100 \)       =>       3336
>   \[4\]     int\( 28\.4 \* 117\.5       \)       =>       3337
> 
> I came across this when calculating tax\-inclusive prices @&#8203; 17\.5%
> 
> I am now using \[4\] instead of \[3\]\, which doesn't appear to barf on any
> values that I throw at it\, but I'd like to know exactly what is
> happening here\.
> 
> I assume that the result of the arithmetic above is 3336\.999999999
> or similar\, and it is rounded by perl because of precision differences\,
> but in \[3\] int\(\) is given the un\-rounded value and returning the integer
> part\. What I don't understand is why \[3\] differs to \[4\]\.

It does indeed relate to rounding\, as you suspected&#8203;:

/usr/bin/perl \-e 'printf "%\.24g\\n"\, 28\.4 \* 117\.5'
3337
/usr/bin/perl \-e 'printf "%\.24g\\n"\, 28\.4 \* 1\.175 \* 100'
3336\.99999999999954525265

\[on x86 perl5\.6\.0 also using doubles\, as yours is\]

I would suspect it is because of the way floating point numbers stored
as \(binary fraction\) \* 2 \*\* n

117\.5 can be precisely represented as 

\(235/256\) \* 2 \*\* 7

\[no approximation\, as 235/256 isn't an infinite binary fraction\]

1\.175 can't be precisely represented; it will be

\(0\.5875\) \* 2 \*\* 1

where the decimal fraction 0\.5875 is an infinite binary fraction

\(cf writing 1/3 as 0\.33333333 in decimal\)

as 1\.175 can't be precisely represented\, the floating point value will either
be slightly too big or slightly too small\. In this case it's too small\,
which is why you see the problem when you use int\.

I'm not sure of the correct generic solution for dealing with this sort
of thing \(in money\)
It may well be "do your calculations in fixed point by using integers\,
handle all rounding explicity"

I think if VAT were lowered to 12\.5% this rounding problem would go away\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From @JohnPeacock

My correct solution is to use Math&#8203;::Currency \(which\, oddly enough\, I 
wrote to deal with exactly this problem\)\.  It does all currency math
using integers \(ultimately\) and does the rounding correctly \(as defined
by the IEEE specs\, i\.e\. even rounding\)\.  It also handles all of your
currency formatting issues and is completely user modifiable and even
locale aware\.

Math&#8203;::Currency relies on Math&#8203;::FixedPrecision to do its work\, but be
warned that you need a patched Math&#8203;::BigFloat and Math&#8203;::BigInt because 
all core Perl versions prior to 5\.6\.1 \(not released\) and 5\.7\.0 have 
problems\.  I have included the new files in the current release on CPAN
\(which I uploaded today\)\.

Get them both from&#8203;:

&nbsp;    http&#8203;://www\.cpan\.org/modules/by\-module/Math/JPEACOCK/

Enjoy

John Peacock

Nicholas Clark wrote&#8203;:
> 
> I'm not sure of the correct generic solution for dealing with this sort
> of thing \(in money\)
> It may well be "do your calculations in fixed point by using integers\,
> handle all rounding explicity"
> 
> I think if VAT were lowered to 12\.5% this rounding problem would go away\.
> 
> Nicholas Clark

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

\-\-\-\-\-BEGIN PGP SIGNED MESSAGE\-\-\-\-\-

Moin\,

Nick wrote&#8203;:
>On Wed\, Feb 21\, 2001 at 05&#8203;:49&#8203;:02PM \+0000\, cdhs@&#8203;tara\.commerce\.uk\.net wrote&#8203;:
>> 
>> This is a bug report for perl from cdhs@&#8203;commerce\.uk\.net\,
>> generated with the help of perlbug 1\.28 running under perl v5\.6\.0\.
>> 
>> 
>> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

>I'm not sure of the correct generic solution for dealing with this sort
>of thing \(in money\) It may well be "do your calculations in fixed point by
> using integers\, handle all rounding explicity"

I think having a look at search\.cpan\.org for Math&#8203;::FixedPrecision may able
to help you ;\)

>I think if VAT were lowered to 12\.5% this rounding problem would go away\.

Come to Germany\, we only have 16% instead of 17\.5\. ;\-\)

Cheers\,

\- \-\- 
perl \-MMath&#8203;::String \-e 'print \\
print Math&#8203;::String\->from\_number\("\+215960156869840440586892398248"\)\,"\\n"'

 http&#8203;://bloodgate\.com/thief/     Thief \- The Dark Project
 http&#8203;://bloodgate\.com/perl       My current Perl projects
 http&#8203;://freedomforlinks\.de       Fight for your right to link\.
 PGP key available on http&#8203;://bloodgate\.com/tels\.asc or via email 

\-\-\-\-\-BEGIN PGP SIGNATURE\-\-\-\-\-
Version&#8203;: 2\.6\.3i
Charset&#8203;: latin1

iQEVAwUBOpVg5ncLPEOTuEwVAQFgnwf9GkSlhXQ9XWkUWbn2HycpOb6t\+qtLSkgF
18zPi9HH2fDkJBuld3UPDa/uHo626DPTjsMCgk8l0IUsn8JhGegr\+KBsw0i1IpK5
/QoMg9LuH1G4h492SiFlcmHG7rdJLUOAlwviOJFyLT\+jmo75CKz6A\+Dz4k1wHQGh
ZZFitb5oIHdScNju2Di22qTqmSRIR5Gu\+LBefNnWwf9shxzFfO3gKR6TqVreJlpg
rNOrhDrtNKCJqTFaSWa9J9MPNC7pFioHmpbsbNSNvngaZIBwOw\+5I\+0/BTbe\+kPo
wimyO7lt6eWbwyBSXSD83HD/qVFtgck1Z\+Wc86\+GhuIZxC61AraGLA==
=kT/a
\-\-\-\-\-END PGP SIGNATURE\-\-\-\-\-
